lib/rouge/lexers/csound.rb

```diff
@@ -0,0 +1,503 @@
+        rule /\bopcode\b/, Keyword::Declaration, :after_opcode_keyword
+        rule /\b(?:end(?:in|op))\b/, Keyword::Declaration
+
+        mixin :partial_statements
+      end
+
+      state :partial_statements do
+        rule /\b(?:0dbfs|A4|k(?:r|smps)|nchnls(?:_i)?|sr)\b/, Name::Variable::Global
+
+        mixin :numbers
+
+        rule %r(\+=|-=|\*=|/=|<<|>>|<=|>=|==|!=|&&|\|\||[~¬]|[=!+\-*/^%&|<>#?:]), Operator
+        rule /[(),\[\]]/, Punctuation
+
+        rule /"/, Str, :quoted_string
+        rule /{{/, Str, :braced_string
+
+        rule /\b(?:do|else(?:if)?|end(?:if|until)|fi|i(?:f|then)|kthen|od|then|until|while)\b/, Keyword
+        rule /\br(?:ir)?eturn\b/, Keyword::Pseudo
+
+        rule /\b[ik]?goto\b/, Keyword, :goto_before_label
+        rule /\b(r(?:einit|igoto)|tigoto)(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+        end
+        rule /\b(c(?:g|in?|k|nk?)goto)(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+          push :goto_before_argument
+        end
+        rule /\b(timout)(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+          push :goto_before_argument
+          push :goto_before_argument
+        end
+        rule /\b(loop_[gl][et])(\(|\b)/ do
+          groups Keyword::Pseudo, Punctuation
+          push :goto_before_label
+          push :goto_before_argument
+          push :goto_before_argument
+          push :goto_before_argument
+        end
+
+        rule /\bprintk?s\b/, Name::Builtin, :prints_opcode
+        rule /\b(?:readscore|scoreline(?:_i)?)\b/, Name::Builtin, :csound_score_opcode
+        rule /\bpyl?run[it]?\b/, Name::Builtin, :python_opcode
+        rule /\blua_(?:exec|opdef)\b/, Name::Builtin, :lua_opcode
+        rule /\bp\d+\b/, Name::Variable::Instance
+        rule /\b(#{Csound.identifier})(?:(:)([A-Za-z]))?\b/ do |m|
+          name = m[1]
+          if self.class.opcodes.include?(name) || self.class.deprecated_opcodes.include?(name)
+            token Name::Builtin, name
+            if m[2]
+              token Punctuation, m[2]
+              token Keyword::Type, m[3]
+            end
+          elsif user_defined_opcodes.include? name
+            token Name::Function, name
+          else
+            if m = /^(g?[aikSw])(\w+)/.match(name)
+              token Keyword::Type, m[1]
+              token Name, m[2]
+            else
+              token Name, name
+            end
+          end
+        end
+      end
+
+      state :instrument_numbers_and_identifiers do
+        mixin :whitespace_and_macro_uses
+        rule /\d+|#{Csound.identifier}/, Name::Function
+        rule /[+,]/, Punctuation
+        rule /\n/, Text, :pop!
+      end
+
+      state :after_opcode_keyword do
+        mixin :whitespace_and_macro_uses
+        rule Csound.identifier do |m|
+          user_defined_opcodes.add m[0]
```

Comment:
A peculiarity of Csound is that the first one or two characters of some (but not all) identifiers are types. For example, `aSignal` is an audio-rate signal, the `a` is a type sigil, so its token should be `Keyword::Type`. (This is what’s going on [here](https://github.com/nwhetsell/rouge/blob/942994248b5402dac905c14016f17b199ce77c55/lib/rouge/lexers/csound.rb#L213-L215).)

However, if someone happens to define an opcode named `aSignal` (or `aGenerator`, or `aSynth`), the `a` no longer indicates a type. The only way to handle this is to maintain some state. Incidentally, this is what the [Pygments Csound lexer](https://bitbucket.org/birkenfeld/pygments-main/src/7941677dc77d4f2bf0bbd6140ade85a9454b8b80/pygments/lexers/csound.py?at=default&fileviewer=file-view-default#csound.py-161) does.